library/blivet.py

```diff
@@ -1265,18 +1289,98 @@ def _manage_encryption(self, members):
 
         return managed_members
 
+    def _manage_thin_pools(self, pool_device):
+
+        existing_thinlvs_obj = [d for d in pool_device.children if d.is_thin_pool]
+
+        existing_thinlvs = list()
+
+        for thinlv_obj in existing_thinlvs_obj:
+            name = thinlv_obj.name.split(self._pool['name'] + '-')[-1]
+            size = thinlv_obj.size
+            existing_thinlvs.append({'name': name, 'size': size})
+
+        thinlvs_to_create = list()
+        new_thinlvs_obj = list()
+
+        # Calculate auto size of thin LVs as follows:
+        # <available space for all autos> = <size of VG> - <20% of VG for thin metadata> - <all user given sizes>
+        # <auto size> = <available space for all autos> / (<count of thinLVs> + <count of regular LVs>)
+        auto_size_dev_count = 0
+        reserved_space = Size(0)
+
+        for volume in self._pool.get('volumes'):
+            if not volume['state']:
+                continue
+
+            if volume['thin']:
+                thin_name = volume.get('thin_pool_name')
+                thin_size = volume.get('thin_pool_size')
+                if thin_name not in [t['name'] for t in thinlvs_to_create] + [d['name'] for d in existing_thinlvs]:
+                    thinlvs_to_create.append({'name': thin_name, 'size': thin_size})
+                    if thin_size is None:
+                        auto_size_dev_count += 1
+                    else:
+                        reserved_space += Size(thin_size)
+            else:
+                # regular LV just take its size
+                vol_size = volume.get('size')
+                if vol_size is None:
+                    auto_size_dev_count += 1
+                else:
+                    reserved_space += Size(vol_size)
```

Comment:
This is causing the wrong error message to be thrown.  The tests are expecting an error from `_get_size()` like `"invalid percentage '2x%' size specified for volume '%s'"`.  Instead, the error message is more cryptic:
```
ValueError: Failed to recognize unit from the spec: 2x%
```
I think you are not supposed to use `Size()` directly unless you are sure the value is not a percentage.  I think you are supposed to use `Size()` directly unless you are sure the value is not a percentage.  I think you are supposed to use `self._get_size()`.  Unfortunately, I'm not sure which object to use for the `_get_size` method - the code is in `class BlivetLVMPool(BlivetPool)` but only `BlivetVolume` has the `_get_size` method.

Alternately you could change the test to look for the above error message, but IMO that isn't a good error message.

Reply:
Thank you. Fixed.